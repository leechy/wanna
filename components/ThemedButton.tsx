import { useThemeColor } from '@/hooks/useThemeColor';

import {
  Text,
  type ViewStyle,
  StyleSheet,
  Pressable,
  Platform,
  View,
  useColorScheme,
} from 'react-native';
import { LinearGradient } from 'expo-linear-gradient';
import { fullButtonBackgrounds } from '@/constants/Colors';

export type ThemedButtonProps = {
  style?: ViewStyle;
  lightColor?: string;
  darkColor?: string;
  title: String;
  onPress?: () => void;
};

export function ThemedButton({
  style,
  lightColor,
  darkColor,
  title,
  onPress,
}: ThemedButtonProps) {
  const fullButtonText = useThemeColor(
    { light: lightColor, dark: darkColor },
    'fullButtonText'
  );
  const theme = useColorScheme() ?? 'light';
  const fullButtonBackground = fullButtonBackgrounds[theme];

  return (
    <View style={[styles.container, style]}>
      <Pressable
        android_ripple={{
          color: fullButtonText,
        }}
        style={({ pressed }) => [
          Platform.select({
            ios: {
              opacity: pressed ? 0.6 : 1,
            },
          }),
        ]}
        onPress={onPress}
      >
        <LinearGradient
          // Button Linear Gradient
          colors={fullButtonBackground as [string, string, ...string[]]}
          locations={[0, 0.23, 1]}
          start={[0.3, -0.3]}
          end={[0.302, 1]}
          style={styles.button}
        >
          <Text
            style={[
              {
                color: fullButtonText,
              },
              styles.text,
            ]}
          >
            {title}
          </Text>
        </LinearGradient>
      </Pressable>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    marginVertical: 12,
    borderRadius: 8,
    shadowColor: '#000',
    boxShadow: '0px 0px 4px 1px rgba(0, 0, 0, 0.25)',
  },
  button: {
    minHeight: 52,
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    paddingHorizontal: 24,
    borderRadius: 8,
  },
  text: {
    fontFamily: 'Nunito',
    fontSize: 21,
    fontWeight: '900',
    lineHeight: 32,
    textAlign: 'center',
  },
});
