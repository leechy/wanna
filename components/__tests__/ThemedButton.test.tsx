import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { ThemedButton } from '../ThemedButton';

describe('ThemedButton', () => {
  it('renders correctly', () => {
    const { getByText } = render(
      <ThemedButton title="Click me" onPress={() => {}} />
    );
    expect(getByText('Click me')).toBeTruthy();
  });

  it('calls onPress when pressed', () => {
    const onPressMock = jest.fn();
    const { getByText } = render(
      <ThemedButton title="Click me" onPress={onPressMock} />
    );
    const button = getByText('Click me');

    fireEvent.press(button);
    expect(onPressMock).toHaveBeenCalled();
  });

  it('matches snapshot', () => {
    const { toJSON } = render(
      <ThemedButton title="Click me" onPress={() => {}} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
