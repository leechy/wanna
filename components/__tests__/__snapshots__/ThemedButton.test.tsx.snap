// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemedButton matches snapshot 1`] = `
<View
  style={
    [
      {
        "alignSelf": "stretch",
        "borderRadius": 8,
        "boxShadow": "0px 0px 4px 1px rgba(0, 0, 0, 0.25)",
        "marginVertical": 12,
        "shadowColor": "#000",
      },
      undefined,
    ]
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    colors={
      [
        4294956800,
        4294961152,
        4294956800,
      ]
    }
    endPoint={
      [
        0.302,
        1,
      ]
    }
    locations={
      [
        0,
        0.23,
        1,
      ]
    }
    startPoint={
      [
        0.3,
        -0.3,
      ]
    }
    style={
      {
        "borderRadius": 8,
      }
    }
  >
    <View
      accessibilityHint="Press to Click me"
      accessibilityLabel="Click me"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderRadius": 8,
            "flexDirection": "row",
            "justifyContent": "center",
            "minHeight": 52,
            "paddingHorizontal": 24,
            "width": "100%",
          },
          {
            "backgroundColor": "transparent",
            "opacity": 1,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000000",
            },
            {
              "fontFamily": "Nunito",
              "fontSize": 21,
              "fontWeight": "900",
              "lineHeight": 32,
              "textAlign": "center",
            },
          ]
        }
      >
        Click me
      </Text>
    </View>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;
